module/zfs/bpobj.c

```diff
@@ -242,119 +277,186 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+		/*
+		 * We've finished with this bpo's directly-referenced BP's.
+		 * If it has no subobjs, we are done processing it. We can free
+		 * its bpobj_info_t (unless it is the topmost, initial_bpo).
+		 * If we are freeing from disk, we can also do that.
+		 */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
 			}
-		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
-		}
-
-		bpobj_close(&sublist);
-		if (err)
-			break;
-		if (free) {
-			err = dmu_object_free(bpo->bpo_os,
-			    objarray[blkoff], tx);
+			/* The initial_bpo has no parent. */
+			if (free && (bpo != initial_bpo)) {
+				bpobj_t *p = bpi->bpi_parent->bpi_bpo;
+
+				ASSERT0(bpo->bpo_phys->bpo_num_blkptrs);
+				ASSERT3U(p->bpo_phys->bpo_num_subobjs, >, 0);
+				ASSERT3U(bpi->bpi_index, ==,
+				    p->bpo_phys->bpo_num_subobjs - 1);
+
+				p->bpo_phys->bpo_num_subobjs--;
+
+				VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
+				    p->bpo_phys->bpo_subobjs,
+				    bpi->bpi_index * sizeof (uint64_t),
+				    sizeof (uint64_t), tx));
+
+				/* eliminate the empty subobj */
+				if (bpo->bpo_havesubobj &&
+				    bpo->bpo_phys->bpo_subobjs != 0) {
+					ASSERT0(bpo->bpo_phys->bpo_num_subobjs);
+					err = dmu_object_free(bpo->bpo_os,
+					    bpo->bpo_phys->bpo_subobjs, tx);
+					if (err) {
+						mutex_exit(&bpo->bpo_lock);
+						break;
+					}
+					dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+					bpo->bpo_phys->bpo_subobjs = 0;
+				}
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
+				if (err) {
+					mutex_exit(&bpo->bpo_lock);
+					break;
+				}
+			}
+			mutex_exit(&bpo->bpo_lock);
+			/* We don't free the inital bpo. */
+			if (bpo != initial_bpo) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+		} else {
+			ASSERT(bpo->bpo_havecomp);
+			/*
+			 * Still need to process subobjs.  Place the bpi back
+			 * in the list ahead of its children.  It will be
+			 * processed again up once the children are all
```

Comment:
I think we actually process (visit) this bpi again after visiting just one child.  At which point if there are more children we'll re-push ourself and push one more child.

typo: extra "up"?